### ISSUE TITLE
PHPUnit 5.3 50% slower than PHPUnit 5.2

### ISSUE THREAD

**Issue by Slamdunk** [sebastianbergmann/phpunit#2128]
Hi, I just updated from 5.2.12 to 5.3.0, and my SUT slowed down by **50%**, from 2 minutes to 3 minutes :open_mouth: 

After some digging, reverting and testing, I found out that this is the commit that introduced the "bug":
https://github.com/sebastianbergmann/phpunit/commit/202600758fa4794ae83b178b919756f69eb2ce59

Let me know if I can provide more feedbacks

**Comment by sebastianbergmann** [comment]
Am I right in assuming that if you experience such a performance degradation that you are using `@depends` a lot to pass fixture around?

**Comment by Slamdunk** [comment]
@sebastianbergmann unfortunately no: we had just two tests linked together with `@depends`, we deleted both code, but the results are the same.

The SUT is a private business application, so I find it hard to extrapolate the source of slowness and public it.

In the next week we are going to write an internal benchmark to better understand if, where and how the are slowing.

In the meantime I can describe out SUT this way:

1. Ubuntu 15.10, PHP 5.6, MySQL 5.6
1. ~ 1.500 tests: 1/3 unit-tests, 1/3 acceptance tests, 1/3 end-to-end tests
1. No Selenium, just pure PHPUnit, no external Mocking libraries
1. ZendFramework & Doctrine full MVC setup on every end-to-end test

If you have other ideas to identify the root of the bug, let us know please.

**Comment by sebastianbergmann** [comment]
The only other feature of PHPUnit affected by the patch you reference is data providers.

**Comment by Slamdunk** [comment]
@sebastianbergmann ok, `@dataProvider` is the source, here a test script:
```php
class MyTest extends PHPUnit_Framework_TestCase
{
    /**
     * @dataProvider dataProvider
     */
    public function testSlowness(stdClass $em)
    {
    }
    
    public function dataProvider()
    {
        $dataProvider = array();
        for ($i = 0; $i < 10; ++$i) {
            $argument = new stdClass();
            for ($j = 0; $j < 1000; ++$j) {
                $argument->{$j} = new stdClass();
            }
            $dataProvider[] = array($argument);
        }
        
        return $dataProvider;
    }
}
```
And the results:
```
$ phpunit52 MyTest.php
PHPUnit 5.2.12 by Sebastian Bergmann and contributors.

..........                                                        10 / 10 (100%)

Time: 25 ms, Memory: 3.25Mb

OK (10 tests, 0 assertions)

$ phpunit53 MyTest.php
PHPUnit 5.3.0 by Sebastian Bergmann and contributors.

..........                                                        10 / 10 (100%)

Time: 336 ms, Memory: 4.50Mb

OK (10 tests, 0 assertions)
```
As you can see the time increases by a 10x factor.

**Comment by sebastianbergmann** [comment]
Not surprising that the slowdown increases with the size of the provided data.

**Comment by JamesGuthrie** [comment]
I have been hit by this as well. One test which is run with a fair amount of data from a provider is now almost unusably slow. On phpunit 5.2 I was able to run a test with all 128 data items in ~25s, on 5.3 I'm seeing it take ~10s per data item.

**Comment by sebastianbergmann** [comment]
The changes made in 202600758fa4794ae83b178b919756f69eb2ce59 are required for mock objects passed into a test using `@dataProvider` or `@depends` to work properly.

Maybe I have to add an option (off by default) to enable the logic added in 202600758fa4794ae83b178b919756f69eb2ce59.

**Comment by bwoebi** [comment]
It's pretty much @dataProvider. It slows down our test suite by over 5 times; taking multiple minutes to complete as opposed to 20 seconds.

Try with https://github.com/amphp/aerys (with 5.3) - it's become so slow and memory intensive (it suddenly needed more than 1 GB of memory instead of 500-600 MB before and thus fatal'ed on travis...).

We're having a very big data provider for this test: https://github.com/amphp/aerys/blob/master/test/HPackTest.php

**Comment by sebastianbergmann** [comment]
@bwoebi I'll work on what I outlined in https://github.com/sebastianbergmann/phpunit/issues/2128#issuecomment-205652486 ASAP.

**Comment by UlrichEckhardt** [comment]
I have two questions on this issue:

1. Is the fix already in any release? I just upgraded from 4.x to 5.4.8 and experienced extreme slowdowns from 10 seconds to 2 minutes. Downgrading to the latest 5.2 fixed those issues.
2. Can anyone summarize what kind of use caused the issue? I honestly don't understand the discussions of the internals above.

**Comment by petrkotek** [comment]
I've run into these issues too.

TL;DR: if you pass **big objects** via `@depends`, you're going to experience a slowdown.

Longer version summing up what i've found out so far:

Basically there was an issue in PHP 4.x that if you passed some mock via `@depends` or `@dataProvider`, its expectation weren't checked.

This issue was fixed in early PHP 5.x, but in the way that it "enumerates" (see https://github.com/sebastianbergmann/object-enumerator/blob/master/src/Enumerator.php#L29) everything in all objects passed via `@depends` and `@dataProvider` (i.e. it inspects everything in the passed object via reflection, so if you're passing big objects, it can take long).

As of ea6ed95, `@dataProvider`, `Enumerator::enumerate()` is not used anymore to check passed objects and uses much cheaper check instead (simple `$testArgument instanceof PHPUnit_Framework_MockObject_MockObject` check), **BUT the problem with usage of `@depends` still remains**.
